### PR TITLE
feat: add meta-transaction rate limiting

### PR DESCRIPTION
**Pull request by opaolini**
# Description

# Testing Instructions

# Checklist

-   [ ] Update [documentation](https://github.com/0xProject/website/blob/development/mdx/api/index.mdx) as needed. **Website Documentation PR:**
-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.

### CODE REVIEW

File: src/handlers/meta_transaction_handlers.ts [hunk 1]
```diff
@@ -232,11 +236,35 @@ export class MetaTransactionHandlers {
...
+                // TODO(oskar) - refactor?
+                if (this._rateLimiter !== undefined) {
+                    const { isAllowed, reason } = await this._rateLimiter.isAllowedAsync(apiKey);
+                    if (!isAllowed) {
+                        const ethereumTxn = await this._metaTransactionService.generatePartialExecuteTransactionEthereumTransactionAsync(
+                            zeroExTransaction,
+                            signature,
+                            protocolFee,
+                        );
+                        res.status(HttpStatus.TOO_MANY_REQUESTS).send({
```
Comment (dekz) [inline src/handlers/meta_transaction_handlers.ts:248]: Todo: we should document that the caller can submit this themselves when rate limited. Janky experience but works better with rfqt then another fresh quote? Or maybe just quote in normal flow. Not sure

File: src/config.ts [hunk 2]
```diff
@@ -204,6 +207,56 @@ export const META_TXN_MAX_GAS_PRICE_GWEI: BigNumber = _.isEmpty(process.env.META
...
+      );
+
+export const META_TXN_ROLLING_RATE_LIMITTER_INTERVAL_UNIT: string | undefined = _.isEmpty(
+    process.env.META_TXN_ROLLING_RATE_LIMITTER_INTERVAL_UNIT,
+)
+    ? undefined
+    : assertEnvVarType(
+          'META_TXN_ROLLING_RATE_LIMITTER_INTERVAL_UNIT',
+          process.env.META_TXN_ROLLING_RATE_LIMITTER_INTERVAL_UNIT,
+          EnvVarType.RateLimitType,
```
Comment (dekz) [inline src/config.ts:257]: IntervalUnit

File: src/types.ts [hunk 3]
```diff
@@ -536,5 +538,31 @@ export interface TransactionWatcherSignerServiceConfig {
     heartbeatIntervalMs: number;
     unstickGasMultiplier: number;
     numBlocksUntilConfirmed: number;
+    rateLimiter?: MetaTransactionRateLimiter;
+}
+
+export interface HttpServiceConfig {
+    HTTP_PORT: number;
```
[the remark this answers is not in the document]
Comment (opaolini) [reply src/types.ts]: refactored the whole shebang

File: src/utils/rate-limiters/meta_transaction_composable_rate_limiter.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        context: MetaTransactionRateLimiterContext,
+    ): Promise<MetaTransactionRateLimiterResponse> {
+        for (const rateLimiter of this._rateLimiters) {
+            const { isAllowed, reason } = await rateLimiter.isAllowedAsync(context);
+            if (!isAllowed) {
+                return { isAllowed, reason };
+            }
+        }
+
+        return { isAllowed: true, reason: '' };
```
Comment (dekz) [inline src/utils/rate-limiters/meta_transaction_composable_rate_limiter.ts:27]: suggestion: avoid empty string in the success case.
Perhaps a union type?
```
type MetaTransactionRateLimiterResponse = MetaTransactionRateLimiterAllowedResponse | MetaTransactionRateLimiterRejectedResponse
```

File: src/utils/rate-limiters/meta_transaction_value_limiter.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    public async isAllowedAsync(
+        context: MetaTransactionRateLimiterContext,
+    ): Promise<MetaTransactionRateLimiterResponse> {
+        const key = this.getKeyFromContextOrThrow(context);
+
+        const { sum } = await this._transactionRepository
+            .createQueryBuilder('tx')
+            .select('SUM( (tx.gas_price * tx.gas_used / 10^18) + ( tx.value / 10^18 ) )', 'sum')
```
Comment (dekz) [inline src/utils/rate-limiters/meta_transaction_value_limiter.ts:34]: should be able to simplify to:
```sql
SUM((tx.gas_used * tx.gas_price + tx.value) * 1e-18)
```

File: test/rate_limiter_test.ts [hunk 6]
```diff
@@ -0,0 +1,252 @@
+import { expect } from '@0x/contracts-test-utils';
```
Comment (dekz) [inline test/rate_limiter_test.ts:1]: we should add a test for `parseUtils.parseJsonStringForMetaTransactionRateLimitConfigOrThrow`

File: src/entities/TransactionEntity.ts [hunk 7]
```diff
@@ -35,10 +35,10 @@ export class TransactionEntity {
     @Column({ name: 'expected_mined_in_sec', type: 'int' })
     public expectedMinedInSec?: number;
 
-    @Column({ name: 'gas_price', type: 'varchar', nullable: true, transformer: BigNumberTransformer })
+    @Column({ name: 'gas_price', type: 'bigint', nullable: true, transformer: BigNumberTransformer })
```
[the remark this answers is not in the document]
Comment (opaolini) [reply src/entities/TransactionEntity.ts:38]: the autosync should handle the fields, I've tried to autogenerate the migration files using `typeorm` CLI but failed.. might take another look. 

But if anything we can test if auto-sync will work before deploying to production, and just create an "init" migration for now.

**Comment by opaolini** [comment]
deploy staging

**Comment by dekz** [comment]
merged https://github.com/0xProject/0x-api/pull/270

**Pull request closed without merging**